File: BehaviorPattern/Strategy/Golang/Concretes/CreditCardStrategy.go
package concretes

import (
	"fmt"
	"v/Interfaces"
)

// Concrete Strategy for Credit Card Payment
type CreditCardStrategy struct {
	cardNumber string
	name string
	cvv string
	dateOfExpiry string
}

// Constructor
func NewCreditCardStrategy(cardNumber string, name string, cvv string, dateOfExpiry string) interfaces.PaymentStrategy {
	return &CreditCardStrategy{
		cardNumber: cardNumber,
		name: name,
		cvv: cvv,
		dateOfExpiry: dateOfExpiry,
	}
}

// Override pay
func (c *CreditCardStrategy) Pay(amount int) {
	fmt.Println(amount, " paid with credit card.")
}

File: BehaviorPattern/Strategy/Java/Concretes/CreditCardStrategy.java
package Concretes;

import Interfaces.PaymentStrategy;

// Concrete Strategy for Credit Card Payment
public class CreditCardStrategy implements PaymentStrategy {
    private String cardNumber;
    private String name;
    private String cvv;
    private String dateOfExpiry;

    public CreditCardStrategy(String cardNumber, String name, String cvv, String dateOfExpiry) {
        this.cardNumber = cardNumber;
        this.name = name;
        this.cvv = cvv;
        this.dateOfExpiry = dateOfExpiry;
    }

    @Override
    public void pay(int amount) {
        System.out.println(amount + " paid with credit card.");
    }
}
